provider/pkg/provider/client/client.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016-2024, Pulumi Corporation.
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/service/cloudcontrol"
+	"github.com/aws/aws-sdk-go-v2/service/cloudcontrol/types"
+	"github.com/mattbaird/jsonpatch"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/schema"
+)
+
+type Client interface {
+	Create(ctx context.Context, typeName string, desiredState map[string]any) (identifier *string, resourceState map[string]any, err error)
+	Read(ctx context.Context, typeName, identifier string) (map[string]interface{}, error)
+	Update(ctx context.Context, typeName, identifier string, patches []jsonpatch.JsonPatchOperation) (map[string]interface{}, error)
+	Delete(ctx context.Context, typeName, identifier string) error
+}
+
+type clientImpl struct {
+	api     CloudControlApiClient
+	awaiter CloudControlAwaiter
+}
+
+func NewClient(cctl *cloudcontrol.Client, roleArn *string) Client {
+	api := NewCloudControlApiClient(cctl, roleArn)
+	return &clientImpl{
+		api:     api,
+		awaiter: NewCloudControlAwaiter(api),
+	}
+}
+
+func (c *clientImpl) Create(ctx context.Context, typeName string, desiredState map[string]any) (identifier *string, resourceState map[string]any, err error) {
+	// Serialize inputs as a desired state JSON.
+	jsonBytes, err := json.Marshal(desiredState)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal as JSON: %w", err)
+	}
+	payload := string(jsonBytes)
+
+	res, err := c.api.CreateResource(ctx, typeName, payload)
+	if err != nil {
+		return nil, nil, fmt.Errorf("creating resource: %w", err)
+	}
+
+	pi, waitErr := c.awaiter.WaitForResourceOpCompletion(ctx, res)
+
+	// Read the state - even if there was a creation error but the progress event contains a resource ID.
+	var id string
+	var outputs map[string]interface{}
+	var readErr error
+	if pi != nil && pi.Identifier != nil {
+		// Retrieve the resource state from AWS.
+		// Note that we do so even if creation hasn't succeeded but the identifier is assigned.
+		id = *pi.Identifier
+		resourceState, err := c.api.GetResource(ctx, typeName, id)
+		if err != nil {
+			readErr = fmt.Errorf("reading resource state: %w", err)
+		} else {
+			outputs = schema.CfnToSdk(resourceState)
+		}
+	}
+
+	if waitErr != nil {
+		if id == "" {
+			return nil, nil, waitErr
+		}
+
+		if readErr != nil {
+			return nil, nil, fmt.Errorf("resource partially created but read failed. read error: %v, create error: %w", readErr, waitErr)
+		}
+
+		// Resource was created but failed to fully initialize.
+		// If it has some state, return a partial error.
+		return &id, outputs, waitErr
```

Comment:
Should we be checking for the id on this branch too so that `id` and `outputs` are both always nill or not nill?

Reply:
I think `outputs` are never nil here? They are assigned if `readErr` is nil, and `readErr` is known to be nil by this line.

So, both `id` and `outputs` are defined here, and we return `nil, nil` everywhere else. What am I missing?